crates/uv-fs/src/lib.rs

```diff
@@ -300,52 +312,55 @@ pub async fn persist_with_retry(
         // So we will update the `from` optional value in safe and borrow-checker friendly way every retry
         // Allows us to use the NamedTempFile inside a FnMut closure used for backoff::retry
         let mut from = Some(from);
-
-        let backoff = backoff_file_move();
-        let persisted = backoff::future::retry(backoff, move || {
+        let persist = move || {
             // Needed because we cannot move out of `from`, a captured variable in an `FnMut` closure, and then pass it to the async move block
-            let mut from = from.take();
+            let mut from: Option<NamedTempFile> = from.take();
 
             async move {
                 if let Some(file) = from.take() {
                     file.persist(to).map_err(|err| {
                         let error_message = err.to_string();
-                        warn!(
-                            "Retrying to persist temporary file to {}: {}",
-                            to.display(),
-                            error_message
-                        );
-
                         // Set back the NamedTempFile returned back by the Error
                         from = Some(err.file);
-
-                        backoff::Error::transient(std::io::Error::new(
-                            std::io::ErrorKind::Other,
-                            format!(
-                                "Failed to persist temporary file to {}: {}",
-                                to.display(),
-                                error_message
-                            ),
-                        ))
+                        PersistRetryError::Persist(error_message)
                     })
                 } else {
-                    Err(backoff::Error::permanent(std::io::Error::new(
-                        std::io::ErrorKind::Other,
-                        format!(
-                            "Failed to retrieve temporary file while trying to persist to {}",
-                            to.display()
-                        ),
-                    )))
+                    Err(PersistRetryError::LostState)
                 }
             }
-        })
-        .await;
+        };
+
+        let persisted = persist
+            .retry(backoff_file_move())
+            .sleep(tokio::time::sleep)
+            .when(|err| matches!(err, PersistRetryError::Persist(_)))
+            .notify(|err, _dur| {
+                if let PersistRetryError::Persist(error_message) = err {
+                    warn!(
```

Comment:
This extra if-let from forgetting the analysis we did to decide to retry is the biggest "flaw" with backon's different API. small potatoes.